src/passes/OptimizeInstructions.cpp

```diff
@@ -458,6 +458,91 @@ struct OptimizeInstructions
           return c;
         }
       }
+      {
+        Const* c;
+        Binary* bin;
+        Expression *x, *y;
+        BinaryOp op;
+
+        // fneg(x) * fneg(y)   ==>   x * y
+        // fneg(x) / fneg(y)   ==>   x / y
+        if (matches(
+              curr,
+              binary(&bin, Mul, unary(Neg, any(&x)), unary(Neg, any(&y)))) ||
+            matches(
+              curr,
+              binary(&bin, DivS, unary(Neg, any(&x)), unary(Neg, any(&y))))) {
+          bin->left = x;
+          bin->right = y;
+          return bin;
+        }
+        // fneg(x * C)   ==>   x * -C
+        // fneg(x / C)   ==>   x / -C
+        // fneg(C / x)   ==>   -C / x
+        //
+        // TODO: using `fastMath` is not necessary, but due to fuzzer can't
+        // flexible handle sign of NaNs we use it here.
+        if (fastMath &&
+            (matches(curr, unary(Neg, binary(&bin, Mul, any(&x), fval(&c)))) ||
+             matches(curr, unary(Neg, binary(&bin, DivS, any(&x), fval(&c)))) ||
+             matches(curr,
+                     unary(Neg, binary(&bin, DivS, fval(&c), any(&x)))))) {
+          c->value = c->value.neg();
+          return bin;
+        }
+        // (fneg(x) * y) op C   ==>   (x * y) op -C
+        // (x * fneg(y)) op C   ==>   (x * y) op -C
+        // (fneg(x) / y) op C   ==>   (x / y) op -C
+        // (x / fneg(y)) op C   ==>   (x / y) op -C
+        //   where op = `*` `/` or `+`
+        if ((matches(curr,
+                     binary(&op,
+                            binary(&bin, Mul, unary(Neg, any(&x)), any(&y)),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, Mul, any(&x), unary(Neg, any(&y))),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, DivS, unary(Neg, any(&x)), any(&y)),
+                            fval(&c))) ||
+             matches(curr,
+                     binary(&op,
+                            binary(&bin, DivS, any(&x), unary(Neg, any(&y))),
+                            fval(&c)))) &&
+            (op == Abstract::getBinary(bin->type, Mul) ||
+             op == Abstract::getBinary(bin->type, DivS) ||
+             op == Abstract::getBinary(bin->type, Add)) &&
+            !x->is<Const>() && !y->is<Const>()) {
```

Comment:
why disallow a Const here? It should be valid with a const as well.

Reply:
if any of `x` or `y`  is constant other rules will be apply:
https://github.com/WebAssembly/binaryen/blob/master/src/passes/OptimizeInstructions.cpp#L1898
https://github.com/WebAssembly/binaryen/blob/master/src/passes/OptimizeInstructions.cpp#L1821

Reply:
This would require a comment, at minimum, then, to explain the reason. But, I hope we don't need it - can we just reorder the rules, putting those other ones before this one? Or is there some really complex ordering problem here? (if so, that should go in the comment)

Reply:
I try to use such order agnostic approach because it more error prone. Also in theory this rules could autogen in future I guess so it will be also required order agnostic. So I'll add comment